packages/toolpad-components/src/DatePicker.tsx

```diff
@@ -0,0 +1,93 @@
+import * as React from 'react';
+
+import { TextField } from '@mui/material';
+
+import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
+import { DesktopDatePicker, DesktopDatePickerProps } from '@mui/x-date-pickers/DesktopDatePicker';
+import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
+import { createComponent } from '@mui/toolpad-core';
+import { Dayjs } from 'dayjs';
+
+export interface Props extends DesktopDatePickerProps<string, Dayjs> {
+  format: string;
+  separator: string;
+  fullWidth: boolean;
+  variant: 'outlined' | 'filled' | 'standard';
+  size: 'small' | 'medium';
+  sx: any;
+}
+
+const resolveFormat = (format: string, separator: string) => format.replaceAll(' ', separator);
+
+function DatePicker(props: Props) {
+  return (
+    <LocalizationProvider dateAdapter={AdapterDayjs as any}>
+      <DesktopDatePicker
+        inputFormat={resolveFormat(props.format, props.separator)}
+        {...props}
+        renderInput={(params) => (
+          <TextField
+            {...params}
+            fullWidth={props.fullWidth}
+            variant={props.variant}
+            size={props.size}
+            sx={props.sx}
+          />
+        )}
+      />
+    </LocalizationProvider>
+  );
+}
+
+export default createComponent(DatePicker, {
+  argTypes: {
+    value: {
+      typeDef: { type: 'string' },
+      onChangeProp: 'onChange',
+      onChangeHandler: (newValue: Dayjs, { format, separator }: Props) => {
+        return newValue.format(resolveFormat(format, separator));
+      },
+      defaultValue: '',
+      defaultValueProp: 'defaultValue',
+    },
+    format: {
```

Comment:
Currently, the Toolpad app developer decides for every user which format they have to input dates in. I wonder if it rather makes more sense that users input the date in the format based on their locale

Reply:
>  wonder if it rather makes more sense that users input the date in the format based on their locale

I'm not really sure how that would work if say 2 users with different locales edited app, which one should be used? Or is there a possibility that DD-MM-YYYY on one machine might get converted to MM-DD-YYYY on the other? Also what if I don't want to use my locale because I'm building an app for users in a different locale?
IMO having an option to choose custom format should be among something that user is able to control, even retool has it:
<img width="764" alt="CleanShot 2023-01-02 at 11 54 07@2x" src="https://user-images.githubusercontent.com/437214/210216165-43d643ac-2145-4b5d-9052-4778c098c961.png">

Reply:
> I'm not really sure how that would work if say 2 users with different locales edited app, which one should be used?

Each user in the locale of their browser. The `value` prop should be kept in an unambiguous format so that it doesn't matter to the runtime.

> Also what if I don't want to use my locale because I'm building an app for users in a different locale?

Those end users will have their browsers set to their own locale, so if the date input just follows that, then it should be fine. What if I build applications for users in mixed locales? e.g. like in MUI: I, in Belgium should see "31/12/2022", and someone in US should see `12-31-2022`, and the `value` prop should be `2022-12-31` for both cases

> IMO having an option to choose custom format should be among something that user is able to control

Which user, the app editor? or the end user? in any case, we can always do both options, default to "use end user locale" and allow overriding with a custom format (and we can probably consolidate the `separator` and `format` props into 1 "format" prop).

Reply:
> The value prop should be kept in an unambiguous format so that it doesn't matter to the runtime.

When I use datePicker.value in a text field somewhere do we display raw value or formatted with locale? If formatted - do we check all values if it matches date string or shall we also use some meta data from datepicker field to know that we are dealing with date value?


> in any case, we can always do both options, default to "use end user locale" and allow overriding with a custom format

Do I understand correctly that you'd like to have `format` input field which is free form text field allowing to pass any format that is supported and if nothing is provided default to locale format?

Reply:
> When I use `datePicker.value` in a text field somewhere do we display raw value or formatted with locale?

I believe we should keep this value unformatted. i.e. a shorthand version of ISO8601 should be fine (`2022-12-31`). I think the most important is that this value is unambiguous, I should be able to do `new Date(datePicker.value)` and that should interpret the correct date everywhere. This is important because often these values will be sent back to the server, and will then lose all the context of the end user locale.
If the user wants to display such a value somewhere in the UI, for now they'll have to expose a format function on the global scope and use that to display the date. Fully aware this has more friction than it could have, but we can iterate on that in a separate effort.

> Do I understand correctly that you'd like to have `format` input field which is free form text field allowing to pass any format that is supported and if nothing is provided default to locale format?

It's an option yes, my preference would be free form text, leave empty to use end user locale, default empty. But we can iterate on the idea.